File: src/code-samples/langchain/agent-invocation-thread-id.ts
// :snippet-start: agent-invocation-thread-id-js
import { AIMessage } from "@langchain/core/messages";
import { createAgent } from "langchain";
import { MemorySaver } from "@langchain/langgraph";

const agent = createAgent({
  model: "claude-sonnet-4-6",
  tools: [],
  checkpointer: new MemorySaver(),
});

const config = { configurable: { thread_id: crypto.randomUUID() } };

let result = await agent.invoke(
  {
    messages: [
      { role: "user", content: "What's the weather in San Francisco?" },
    ],
  },
  config,
);

// A follow-up turn on the same conversation: reuse the same thread_id to keep history
result = await agent.invoke(
  { messages: [{ role: "user", content: "What about tomorrow?" }] },
  config,
);
// :snippet-end:

// :remove-start:
function lastAssistantText(): string {
  const last = result.messages[result.messages.length - 1];
  if (!(last instanceof AIMessage)) {
    throw new Error("expected final assistant message");
  }
  const blocks = last.contentBlocks ?? [];
  const textBlock = blocks.find((b) => b.type === "text");
  if (textBlock && textBlock.type === "text") {
    return textBlock.text;
  }
  return typeof last.content === "string" ? last.content : "";
}

async function main() {
  const humanCount = result.messages.filter((m) => m.type === "human").length;
  if (humanCount < 2) {
    throw new Error(`expected >=2 human turns, got ${humanCount}`);
  }
  const aiCount = result.messages.filter((m) => m.type === "ai").length;
  if (aiCount < 2) {
    throw new Error(`expected >=2 assistant replies, got ${aiCount}`);
  }
  if (!lastAssistantText()) {
    throw new Error("expected non-empty final assistant text");
  }
  console.log("✓ thread_id invocation persists conversation across turns");
}

main();
// :remove-end:

File: src/code-samples/langchain/agent-invocation-thread-id.py
# :snippet-start: agent-invocation-thread-id-py
from langchain.agents import create_agent
from langchain_core.utils.uuid import uuid7
from langgraph.checkpoint.memory import InMemorySaver

agent = create_agent(
    model="claude-sonnet-4-6",
    tools=[],
    checkpointer=InMemorySaver(),
)

config = {"configurable": {"thread_id": str(uuid7())}}

result = agent.invoke(
    {"messages": [{"role": "user", "content": "What's the weather in San Francisco?"}]},
    config=config,
)

# A follow-up turn on the same conversation: reuse the same thread_id to keep history
result = agent.invoke(
    {"messages": [{"role": "user", "content": "What about tomorrow?"}]},
    config=config,
)
# :snippet-end:

# :remove-start:
if __name__ == "__main__":
    last = result["messages"][-1]
    text = last.content_blocks[0]["text"] if last.content_blocks else str(last.content)
    assert text, "expected non-empty final assistant text"
    human_msgs = [m for m in result["messages"] if m.type == "human"]
    assert len(human_msgs) >= 2
    ai_msgs = [m for m in result["messages"] if m.type == "ai"]
    assert len(ai_msgs) >= 2

    print("✓ thread_id invocation persists conversation across turns")
# :remove-end:
